synda/model/run.py

```diff
@@ -55,19 +55,27 @@ def create_with_steps(session: Session, config: "Config") -> "Run":
         return run
 
     @staticmethod
-    def restart_from_step(session: Session, config: "Config", step: "Step") -> tuple["Run", list[Node], list[Step]]:
+    def restart_from_step(
+        session: Session, config: "Config", step: "Step"
+    ) -> tuple["Run", list[Node], list[Step]]:
         run = session.exec(select(Run).where(Run.id == step.run_id)).first()
         if run.config != config.model_dump():
-            raise ValidationError("The actual config is different from the restarted run config")
+            raise ValidationError(
+                "The actual config is different from the restarted run config"
+            )
         run.update(session, RunStatus.RUNNING)
 
         input_node_ids: list[StepNode] = session.exec(
-            select(StepNode.node_id).where(StepNode.step_id==step.id)
+            select(StepNode.node_id).where(StepNode.step_id == step.id)
+        ).fetchall()
+        input_nodes: list[Node] = session.exec(
+            select(Node).where(Node.id.in_(input_node_ids))
         ).fetchall()
-        input_nodes: list[Node] = session.exec(select(Node).where(Node.id.in_(input_node_ids))).fetchall()
-        return run, input_nodes, run.steps[step.position-1:]
+        return run, input_nodes, run.steps[step.position - 1 :]
```

Comment:
Can be optimized with a single query with a JOIN statement. 

(and should be moved to the Node model)